### PR TITLE
Unload plugins

### PR DESCRIPTION
**Pull request by changlan**
- Use static object's constructor and destructor to initialize and destruct a module class.
- Check if workers are running before importing/unloading plugins.
- Misc. fixes (renaming, comments, etc.)

### CODE REVIEW

**Comment by changlan** [comment]
Not ready yet. Libraries compiled with GCC do not execute destructors upon `dlclose()`.

File: core/bessd.cc [hunk 1]
```diff
@@ -368,11 +368,32 @@ static inline bool HasSuffix(const std::string &s, const std::string &suffix) {
          std::equal(suffix.rbegin(), suffix.rend(), s.rbegin());
 }
 
-bool LoadModule(const std::string &module_path) {
-  return (dlopen(module_path.c_str(), RTLD_NOW) != nullptr);
+static std::unordered_map<std::string, void *> plugin_handles;
```
Comment (barath) [inline core/bessd.cc]: Could you add a comment on key / values for this.

File: core/bessd.cc [hunk 2]
```diff
@@ -368,11 +368,32 @@ static inline bool HasSuffix(const std::string &s, const std::string &suffix) {
...
+
+bool LoadPlugin(const std::string &path) {
+  void *handle = dlopen(path.c_str(), RTLD_NOW);
+  if (handle != nullptr) {
+    plugin_handles.emplace(path, handle);
+    return true;
+  } else {
```
Comment (barath) [inline core/bessd.cc]: No need for else.

File: core/bessd.cc [hunk 3]
```diff
@@ -368,11 +368,32 @@ static inline bool HasSuffix(const std::string &s, const std::string &suffix) {
...
+  }
+}
+
+bool UnloadPlugin(const std::string &path) {
+  auto it = plugin_handles.find(path);
+  if (it == plugin_handles.end()) {
+    VLOG(1) << "Plugin " << path << " not found.";
+    return false;
+  }
+  bool result = (dlclose(it->second) == 0);
```
Comment (barath) [inline core/bessd.cc]: Minor -- maybe rename `result` to `ok` or `success`?

**Comment by barath** [comment]
Also, not totally sure, but this might need to be rebased due to the TC merge.

**Comment by sangjinhan** [comment]
Is this the issue the `-fno-gnu-unique` option trying to address? (adding here for future reference) https://www.redhat.com/archives/posix-c++-wg/2009-August/msg00002.html

What the article says "static class members" (it seems we don't have static variables in inline functions) are the offending variables which prevents from unloading? These are all I found such variables:
* `static std::map<std::string, Module *> all_modules_` (class Module)
* `static std::map<std::string, Port *> all_ports_` (class Port)
* `static std::unordered_map<std::string, TrafficClass *> all_tcs_` (class TrafficClass)

Ideally module plugins should never have a reference to those symbols... Perhaps we should separate `class Module` (what plugins implement) from `class ModuleWrapper` (internally used by the BESS framework) in the long term.

**Comment by changlan** [comment]
Good point. Making an issue so that we can keep track.

**Pull request merged**